db/schema_tables.cc

```diff
@@ -1131,10 +1131,28 @@ static std::vector<V> get_list(const query::result_set_row& row, const sstring&
 // Create types for a given keyspace. This takes care of topologically sorting user defined types.
 template <typename T> static std::vector<user_type> create_types(keyspace_metadata& ks, T&& range) {
     cql_type_parser::raw_builder builder(ks);
+    std::vector<bytes> names;
     for (const query::result_set_row& row : range) {
-        builder.add(row.get_nonnull<sstring>("type_name"),
-                        get_list<sstring>(row, "field_names"),
-                        get_list<sstring>(row, "field_types"));
+        auto name = row.get_nonnull<sstring>("type_name");
+        names.push_back(to_bytes(name));
+        builder.add(std::move(name), get_list<sstring>(row, "field_names"), get_list<sstring>(row, "field_types"));
+    }
+    // Add user types that use any of the above types. From the
+    // database point of view they haven't changed since the content
+    // of system.types is the same for them. The runtime objects in
+    // the other hand now point to out of date types, so we need to
+    // recreate them.
+    for (const auto& p : ks.user_types()->get_all_types()) {
+        const auto&t = p.second;
+        for (const auto& name : names) {
+            if (t->_name != name && t->references_user_type(t->_keyspace, name)) {
+                std::vector<sstring> field_types;
+                for (const data_type& f : t->field_types()) {
+                    field_types.push_back(f->as_cql3_type().to_string());
+                }
+                builder.add(t->get_name_as_string(), t->string_field_names(), std::move(field_types));
```

Comment:
What if `t` is also in `names` (meaning it's being modified here), wouldn't we add its entry twice, with a stale field names?

Also, since we changed `t->_name`, shouldn't we also update types which reference `t->_name` ?

Reply:
We do check if "t->_name != name", so I don't think we add it twice, is that the question?

For the second part, references_user_type checks the transitive property, so if we modified A, B references A and C references B, C->references_user_type(A) returns true.

Reply:
If `names` has more than one element, `t->_name != name` doesn't guarantee that `t` is not in `names`, it could be that `t->_name == name` for some later value of `name`. 

Suppose the type A references B before, and both A and B were changed. Suppose names = [A, B]. You first add A because it was changed, with the new field definitions. Then when you iterate over all types, you add it again because it references B.

It may also happen that when you add A with the new field definitions, you reference the old definition of B because it was not added yet.